### ISSUE TITLE
[Bug] PR #11304 sets top_k = 50 by default which is wrong

### ISSUE THREAD

**Issue by voipmonitor** [sgl-project/sglang#11574]
### Checklist

- [x] 1. I have searched related issues but cannot get the expected help.
- [x] 2. The bug has not been fixed in the latest version.
- [x] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.
- [x] 4. If the issue you raised is not a bug but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
- [x] 5. Please use English, otherwise it will be closed.

### Describe the bug

#11304 changes the default top_k to 50 which might be wrong ? 

[2025-10-14 02:01:37] Using default chat sampling params from model generation config: {'repetition_penalty': 1.0, 'temperature': 1.0, 'top_k': 50, 'top_p': 1.0}

this is how I'm loading model: 

 NCCL_P2P_LEVEL=4 NCCL_DEBUG=INFO CUDA_VISIBLE_DEVICES=4,5,6,7 USE_TRITON_W8A8_FP8_KERNEL=1 SGL_ENABLE_JIT_DEEPGEMM=false PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True python -m sglang.launch_server --model /mnt/GLM-4.5-Air-FP8/  --tp 4 --speculative-algorithm EAGLE   --speculative-num-steps 3   --speculative-eagle-topk 1   --speculative-num-draft-tokens 4 --host 0.0.0.0 --port  4999 --mem-fraction-static 0.80  --context-length 128000  --enable-metrics  --attention-backend flashinfer  --tool-call-parser glm45    --reasoning-parser glm45   --served-model-name glm-4.5-air   --chunked-prefill-size 64736  --enable-mixed-chunk   --cuda-graph-max-bs 1024 --model-loader-extra-config '{"enable_multithread_load": true, "num_threads": 8}'  --sampling-defaults model

there is no top_k configuration in the glm which should set this to 50 - where this 50 is coming from? 

The side effect is that setting top_k to anything else than -1 causes token / sec speed drop by 30% when the eagle speculative is enabled 

### Reproduction

NCCL_P2P_LEVEL=4 NCCL_DEBUG=INFO CUDA_VISIBLE_DEVICES=4,5,6,7 USE_TRITON_W8A8_FP8_KERNEL=1 SGL_ENABLE_JIT_DEEPGEMM=false PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True python -m sglang.launch_server --model /mnt/GLM-4.5-Air-FP8/  --tp 4 --speculative-algorithm EAGLE   --speculative-num-steps 3   --speculative-eagle-topk 1   --speculative-num-draft-tokens 4 --host 0.0.0.0 --port  4999 --mem-fraction-static 0.80  --context-length 128000  --enable-metrics  --attention-backend flashinfer  --tool-call-parser glm45    --reasoning-parser glm45   --served-model-name glm-4.5-air   --chunked-prefill-size 64736  --enable-mixed-chunk   --cuda-graph-max-bs 1024 --model-loader-extra-config '{"enable_multithread_load": true, "num_threads": 8}'  --sampling-defaults model

[2025-10-14 02:01:37] Using default chat sampling params from model generation config: {'repetition_penalty': 1.0, 'temperature': 1.0, 'top_k': 50, 'top_p': 1.0}

### Environment

4x RTX 6000 PRO 
the latest sglang docker

**Comment by adarshxs** [comment]
This is because top_k isn't mentioned specifically in the config for GLM 4.5 but is rather calculated as 
`top_k = num_experts_per_tok + num_fused_shared_experts`

if it isnt set transformers defaults to 50: https://huggingface.co/docs/transformers/main/en/main_classes/text_generation#transformers.GenerationConfig.top_k

will fix this soon

**Comment by voipmonitor** [comment]
> This is because top_k isn't mentioned specifically in the config for GLM 4.5 but is rather calculated as `top_k = num_experts_per_tok + num_fused_shared_experts`
> 
> if it isnt set transformers defaults to 50: https://huggingface.co/docs/transformers/main/en/main_classes/text_generation#transformers.GenerationConfig.top_k
> 
> will fix this soon

is there any consensus / recommendation what GLM should even use? 

Should I also create another issue for slowing down the eagle speculative inference when top_k is set ?

**Comment by adarshxs** [comment]
@voipmonitor what i can see in the [GLM modelling file](https://github.com/sgl-project/sglang/blob/34e2ec886a772f5a109559390bdd2429f0d99fa4/python/sglang/srt/models/glm4_moe.py#L422)

If shared-experts fusion is enabled (default on a single-GPU NV platform with CUDA capability ≥ 8.0 and no expert parallelism), then num_fused_shared_experts = 1 and `top_k` = 8+1 = 9.

If fusion is disabled (e.g., CPU, AMD/HIP, CUDA < 8.0, expert parallel size > 1, or fusion explicitly disabled), then num_fused_shared_experts = 0 and `top_k` = 8.

I've raised a PR for this issue to be fixed.

**Comment by adarshxs** [comment]
> Should I also create another issue for slowing down the eagle speculative inference when top_k is set ?

the reason `top_k` is slowing down the spec inference is because when top_k is set `!=-1` it calls `torch.top_k` filtering on logits each step. This causes a partial sort. cc @merrymercy

**Comment by voipmonitor** [comment]
> > Should I also create another issue for slowing down the eagle speculative inference when top_k is set ?
> 
> the reason `top_k` is slowing down the spec inference is because when top_k is set `!=-1` it calls `torch.top_k` filtering on logits each step. This causes a partial sort. cc [@merrymercy](https://github.com/merrymercy)

why it happens only when speculative is enabled? is it bybdesign / fixable? Speculative inference makes no sense in case topk is used (if it defaults to 50 )
